common/persistence/visibility/store/sql/visibility_store.go

```diff
@@ -647,67 +776,154 @@ func (s *VisibilityStore) rowToInfo(
 	return info, nil
 }
 
-func (s *VisibilityStore) processRowSearchAttributes(
+func (s *VisibilityStore) processRowSearchAttributesSeparated(
 	rowSearchAttributes sqlplugin.VisibilitySearchAttributes,
 	nsName namespace.Name,
 	chasmMapper *chasm.VisibilitySearchAttributesMapper,
-) (*commonpb.SearchAttributes, error) {
+) (map[string]interface{}, *commonpb.SearchAttributes, error) {
 	saTypeMap, err := s.searchAttributesProvider.GetSearchAttributes(
 		s.GetIndexName(),
 		false,
 	)
 	if err != nil {
-		return nil, serviceerror.NewUnavailable(
+		return nil, nil, serviceerror.NewUnavailable(
 			fmt.Sprintf("Unable to read search attributes types: %v", err))
 	}
 
-	combinedTypeMap := make(map[string]enumspb.IndexedValueType)
-	for k, v := range saTypeMap.Custom() {
-		combinedTypeMap[k] = v
-	}
+	// Get registered CHASM SAs for this archetype (if it's a CHASM query)
+	var chasmTypeMap map[string]enumspb.IndexedValueType
 	if chasmMapper != nil {
-		chasmTypeMap := chasmMapper.SATypeMap()
-		for fieldName, fieldType := range chasmTypeMap {
-			combinedTypeMap[fieldName] = fieldType
-		}
+		chasmTypeMap = chasmMapper.SATypeMap()
 	}
-	finalTypeMap := searchattribute.NewNameTypeMap(combinedTypeMap)
 
-	// In SQLite, keyword list can return a string when there's only one element.
-	// This changes it into a slice.
+	customTypeMap := saTypeMap.Custom()
+
+	// Split row search attributes into CHASM and custom based on field name pattern
+	chasmRowSAs := make(sqlplugin.VisibilitySearchAttributes)
+	customRowSAs := make(sqlplugin.VisibilitySearchAttributes)
+
 	for name, value := range rowSearchAttributes {
-		tp, err := finalTypeMap.GetType(name)
-		if err != nil {
-			return nil, err
+		if defs.IsChasmSearchAttribute(name) {
+			// For CHASM queries, only include registered CHASM SAs
+			if chasmMapper != nil {
+				if _, isRegistered := chasmTypeMap[name]; isRegistered {
```

Comment:
Since you are doing further checks below when you aliasing, you don't need to do any checks here, and do all checks together below.